sdk/core/Microsoft.Azure.Core.NewtonsoftJson/tests/NewtonsoftJsonObjectSerializerTest.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Serialization;
+using NUnit.Framework;
+
+namespace Azure.Core.Tests
+{
+    public class NewtonsoftJsonObjectSerializerTest
+    {
+        private static readonly NewtonsoftJsonObjectSerializer JsonObjectSerializer = new NewtonsoftJsonObjectSerializer(true, new JsonSerializerSettings()
+        {
+            ContractResolver = new CamelCasePropertyNamesContractResolver(),
+            Converters = new[]
+            {
+                new StringEnumConverter(true),
+            },
+        });
+
+        [Test]
+        public void CanSerializeAnObject()
+        {
+            using var memoryStream = new MemoryStream();
+            var o = new Model {A = "1", ActuallyB = 2, Type = ModelType.One};
+
+            JsonObjectSerializer.Serialize(memoryStream, o, o.GetType(), default);
+
+            Assert.AreEqual("{\"a\":\"1\",\"b\":2,\"type\":\"one\"}", Encoding.UTF8.GetString(memoryStream.ToArray()));
+        }
+
+        [Test]
+        public async Task CanSerializeAnObjectAsync()
+        {
+            using var memoryStream = new MemoryStream();
+            var o = new Model {A = "1", ActuallyB = 2, Type = ModelType.One};
+
+            await JsonObjectSerializer.SerializeAsync(memoryStream, o, o.GetType(), default);
+
+            Assert.AreEqual("{\"a\":\"1\",\"b\":2,\"type\":\"one\"}", Encoding.UTF8.GetString(memoryStream.ToArray()));
+        }
+
+        [Test]
+        public void CanDeserializeAnObject()
+        {
+            using var memoryStream = new MemoryStream(Encoding.UTF8.GetBytes("{\"a\":\"1\",\"b\":2,\"type\":\"two\"}"));
+
+            var model = (Model)JsonObjectSerializer.Deserialize(memoryStream, typeof(Model), default);
+
+            Assert.AreEqual("1", model.A);
+            Assert.AreEqual(2, model.ActuallyB);
+            Assert.AreEqual(ModelType.Two, model.Type);
+        }
+
+        [Test]
+        public async Task CanDeserializeAnObjectAsync()
+        {
+            using var memoryStream = new MemoryStream(Encoding.UTF8.GetBytes("{\"a\":\"1\",\"b\":2,\"type\":\"two\"}"));
+
+            var model = (Model)await JsonObjectSerializer.DeserializeAsync(memoryStream, typeof(Model), default).ConfigureAwait(false);
+
+            Assert.AreEqual("1", model.A);
+            Assert.AreEqual(2, model.ActuallyB);
+            Assert.AreEqual(ModelType.Two, model.Type);
+        }
+
+        [Test]
+        public void GetSerializedMemberName([Values] bool camelCase)
```

Comment:
As part of asserting that `GetSerializedMemberName` returns the right value in these tests can we run an  actual serialization and verify the property is there/not there. This might be cleaner than relying on indirect sources of the information.

Reply:
After changing the tests, I think we actually need both. This is a good idea, but I also want to make sure that we return null if a property won't be de/serialized. The interesting case with both of these tests is `ReadOnlyD`, which will serialize but not deserialize. I think serialization is the scenario more prone to error (e.g. serializing and sending a document with an extra property to an index errs in the Search Service it sending to the index itself), so "discoverying" the serialized property "d" is more important that ignoring it, like the modified test now does.